clang/labs/lab1/vyunov_danila/AlwaysInlinePlugin.cpp

```diff
@@ -0,0 +1,72 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+
+class AddAlwaysInlineConsumer : public clang::ASTConsumer {
+public:
+  bool hasCondition(clang::Stmt *S) {
+    if (!S) {
+      return false;
+    }
+
+    if (llvm::isa<clang::IfStmt>(S) || llvm::isa<clang::SwitchStmt>(S) ||
+        llvm::isa<clang::WhileStmt>(S) || llvm::isa<clang::DoStmt>(S) ||
+        llvm::isa<clang::ForStmt>(S)) {
+      return true;
+    }
+
+    for (clang::Stmt *Child : S->children()) {
+      if (hasCondition(Child)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  bool HandleTopLevelDecl(clang::DeclGroupRef D) override {
+    clang::FunctionDecl *FD = nullptr;
+
+    for (clang::Decl *Decl : D) {
+      FD = clang::dyn_cast<clang::FunctionDecl>(Decl);
+      if (FD) {
+        if (!FD->hasAttr<clang::AlwaysInlineAttr>() &&
+            !hasCondition(FD->getBody())) {
+          FD->addAttr(
+              clang::AlwaysInlineAttr::CreateImplicit(FD->getASTContext()));
+          FD->print(llvm::outs() << "__attribute__((always_inline)) ");
+        } else
+          FD->print(llvm::outs());
```

Comment:
This is a cheat)
Please check attribute presence/absence in output of `clang -ast-dump ...` command. Not in logs of your plugin.
```suggestion
```

Reply:
Gentle reminder to remove printing of a function into a log and check functions and attributes in a AST dump of a function performed by `clang -ast-dump` option.